File: 57-insert-interval/insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> ans;
        int i = 0;
        while (i<intervals.size() && intervals[i][1]<newInterval[0]){
            ans.push_back(intervals[i]);
            i+=1;
        }

        while (i<intervals.size() && intervals[i][0]<=newInterval[1]){
            newInterval = {min(newInterval[0], intervals[i][0]), max(newInterval[1], intervals[i][1])};
            i+=1;
        }
        ans.push_back(newInterval);

        while (i<intervals.size()){
            ans.push_back(intervals[i]);
            i+=1;
        }

        return ans;
    }
};

File: 57-insert-interval/insert-interval.go
func insert(intervals [][]int, newInterval []int) [][]int {
    var ans [][]int
    i := 0
    for i < len(intervals) && intervals[i][1] < newInterval[0] {
        ans = append(ans, intervals[i])
        i++
    }

    for i < len(intervals) && intervals[i][0] <= newInterval[1] {
        start := min(intervals[i][0], newInterval[0])
        end := max(intervals[i][1], newInterval[1])

        newInterval = []int{start, end}
        i+=1
    }

    ans = append(ans, newInterval)

    for i < len(intervals) {
        ans = append(ans, intervals[i])
        i++
    }
    
    return ans
}